### PR TITLE
perf(Event): Use pooled event objects

### LINKED ISSUE #1212: Add pooling for JS Events
The [Event](https://github.com/Microsoft/react-native-windows/blob/master/ReactWindows/ReactNative.Shared/UIManager/Events/Event.cs) base class should be configured to support pooling, and we should use pooling, especially for high frequency events like mouseOver, onLayout, etc.

### PR DESCRIPTION
**Pull request by rozele**
ReactAndroid uses pooling on scroll, touch, and layout events. Adding the same behavior here.

Fixes #1212

### CODE REVIEW

File: ReactWindows/ReactNative.Net46/Touch/TouchHandler.cs [hunk 1]
```diff
@@ -415,21 +424,54 @@ public override void Dispatch(RCTEventEmitter eventEmitter)
...
+
+            protected override void OnDispose()
+            {
+                base.OnDispose();
+                s_eventsPool.Free(this);
+            }
+
+            private void Init(TouchEventType touchEventType, JArray touches, JArray changedIndices, uint coalescingKey)
+            {
+                Init(-1);
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Touch/TouchHandler.cs]: what's the -1 magic number mean? can it be extracted to a constant?

File: ReactWindows/ReactNative.Net46/Touch/TouchHandler.cs [hunk 2]
```diff
@@ -415,21 +424,54 @@ public override void Dispatch(RCTEventEmitter eventEmitter)
...
+                @event.Init(touchEventType, touches, changedIndices, coalescingKey);
+                return @event;
+            }
         }
 
         class PointerEnterExitEvent : Event
         {
-            private readonly TouchEventType _touchEventType;
+            private static readonly ConcurrentObjectPool<PointerEnterExitEvent> s_eventsPool =
+                new ConcurrentObjectPool<PointerEnterExitEvent>(() => new PointerEnterExitEvent(), 3);
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Touch/TouchHandler.cs]: magic number 3 needs to be extracted to readonly/const

File: ReactWindows/ReactNative.Net46/Touch/TouchHandler.cs [hunk 3]
```diff
@@ -439,14 +481,13 @@ public override void Dispatch(RCTEventEmitter eventEmitter)
                 };
 
                 var enterLeaveEventName = default(string);
-                switch (_touchEventType)
+                if (_touchEventType == TouchEventType.Entered)
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Touch/TouchHandler.cs]: why change this from a switch?

File: ReactWindows/ReactNative.Net46/Views/Scroll/ReactScrollViewManager.cs [hunk 4]
```diff
@@ -496,15 +497,13 @@ private static void ScrollTo(ScrollViewer scrollView, double x, double y, bool a
 
         class ScrollEvent : Event
         {
-            private readonly ScrollEventType _type;
-            private readonly JObject _data;
+            private static readonly ConcurrentObjectPool<ScrollEvent> s_eventsPool =
+                new ConcurrentObjectPool<ScrollEvent>(() => new ScrollEvent(), 3);
```
Comment (matthargett) [inline ReactWindows/ReactNative.Net46/Views/Scroll/ReactScrollViewManager.cs]: magic number 3 needs const/readonly

**Comment by rozele** [comment]
If we're going to pool events, we should update the touch events to hold more fields rather than just the JToken, to ensure that the JToken is only created at the time is it dispatched to JavaScript (making it less likely to end up in Gen1)

**Comment by rozele** [comment]
Closing, not much impact on perf.

**Pull request closed without merging**